### ISSUE TITLE
Load balancer endpoint certificates are uploaded to an incorrect endpoint

### ISSUE THREAD

**Issue by Harppi** [ansible-collections/netapp.storagegrid#5]
##### SUMMARY
An Ansible module called [na_sg_grid_gateway.py](https://github.com/ansible-collections/netapp.storagegrid/blob/main/plugins/modules/na_sg_grid_gateway.py) uploads a load balancer endpoint certificate to an incorrect load balancer endpoint if there are multiple endpoints that use the same port.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

na_sg_grid_gateway

##### ANSIBLE VERSION

```paste below
ansible [core 2.11.6]
  config file = /Users/Ville.Harmaala/.ansible.cfg
  configured module search path = ['/Users/Ville.Harmaala/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/4.8.0/libexec/lib/python3.10/site-packages/ansible
  ansible collection location = /Users/Ville.Harmaala/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.10.0 (default, Oct 13 2021, 06:45:00) [Clang 13.0.0 (clang-1300.0.29.3)]
  jinja version = 3.0.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /Users/Ville.Harmaala/.ansible/collections/ansible_collections
Collection         Version
------------------ -------
netapp.storagegrid 21.8.0
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/Users/Ville.Harmaala/.ansible.cfg) = True
CALLBACKS_ENABLED(/Users/Ville.Harmaala/.ansible.cfg) = ['profile_tasks']
DEFAULT_MANAGED_STR(/Users/Ville.Harmaala/.ansible.cfg) = Ansible managed: modified on %Y-%m-%d %H:%M:%S by {uid} on {host}
DEFAULT_ROLES_PATH(/Users/Ville.Harmaala/.ansible.cfg) = ['<snip>']
HOST_KEY_CHECKING(/Users/Ville.Harmaala/.ansible.cfg) = False
RETRY_FILES_ENABLED(/Users/Ville.Harmaala/.ansible.cfg) = False
```

##### OS / ENVIRONMENT

StorageGRID version: 11.5.0.1-20210615.1852.af39a66
Number of load balancer endpoints: 5
Load balancer endpoint port: 10443

##### STEPS TO REPRODUCE

1. Create two load balancer endpoints called **foo** and **bar**
2. Configure both load balancer endpoints to use port 10443
3. Upload a load balancer endpoint certificate to the endpoint called **foo** by using the Ansible module called [na_sg_grid_gateway.py](https://github.com/ansible-collections/netapp.storagegrid/blob/main/plugins/modules/na_sg_grid_gateway.py)

```yaml
# Authorize
- name: Get grid authorization token
  uri:
    url: "{{ grid_admin_base_url }}/api/v3/authorize"
    method: POST
    body: {
      "username": "{{ grid_username }}",
      "password": "{{ grid_password }}",
      "cookie": false,
      "csrfToken": false
    }
    body_format: json
  register: auth
  
# Deploy a certificate
- name: Deploy load balancer endpoint certificate
  na_sg_grid_gateway:
    api_url: "{{ grid_admin_base_url }}"
    auth_token: "{{ auth.json.data }}"
    state: present
    display_name: "foo"
    port: 10443
    secure: true
    enable_ipv4: true
    enable_ipv6: true
    default_service_type: "s3"
    server_certificate: "<snip>"
    private_key: "<snip>"
    ca_bundle: "<snip>"
  no_log: true  # hide cert, chain and privkey
```

##### EXPECTED RESULTS

The certificate is uploaded to the load balancer endpoint called **foo**.

##### ACTUAL RESULTS

The certificate is actually deployed to the load balancer endpoint called **bar**.

```paste below
changed: [localhost] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "api_url": "<snip>",
            "auth_token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ca_bundle": "<snip>",
            "default_service_type": "s3",
            "display_name": "foo,
            "enable_ipv4": true,
            "enable_ipv6": true,
            "port": 10443,
            "private_key": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "secure": true,
            "server_certificate": "<snip>"
            "state": "present",
            "validate_certs": false
        }
    },
    "msg": "Load Balancer Gateway Port Updated",
    "resp": {
        "accountId": "0",
        "certSource": "plaintext",
        "defaultServiceType": "s3",
        "displayName": "bar",
        "enableIPv4": true,
        "enableIPv6": true,
        "id": "51f0fffd-9264-4b48-baa7-33e21cdd8e3c",
        "pinTargets": {
            "haGroups": [
                "dc22631c-7d36-41a7-bad2-9fdfe472a747"
            ],
            "nodeInterfaces": []
        },
        "plaintextCertData": {
            "caBundleEncoded": "<snip>",
            "metadata": {
                "caBundleDetails": [
                    {
                        "fingerPrints": {
                            "SHA-1": "<snip>",
                            "SHA-256": "<snip>"
                        },
                        "issuer": "/C=US/O=Internet Security Research Group/CN=ISRG Root X1",
                        "notAfter": "2025-09-15T16:00:00.000Z",
                        "notBefore": "2020-09-04T00:00:00.000Z",
                        "serialNumber": "<snip>"
                        "subject": "/C=US/O=Let's Encrypt/CN=R3"
                    },
                    {
                        "fingerPrints": {
                            "SHA-1": "<snip>",
                            "SHA-256": "<snip>"
                        },
                        "issuer": "/O=Digital Signature Trust Co./CN=DST Root CA X3",
                        "notAfter": "2024-09-30T18:14:03.000Z",
                        "notBefore": "2021-01-20T19:14:03.000Z",
                        "serialNumber": "<snip>",
                        "subject": "/C=US/O=Internet Security Research Group/CN=ISRG Root X1"
                    }
                ],
                "serverCertificateDetails": {
                    "fingerPrints": {
                        "SHA-1": "<snip>",
                        "SHA-256": "<snip>"
                    },
                    "issuer": "/C=US/O=Let's Encrypt/CN=R3",
                    "notAfter": "2022-03-02T11:35:53.000Z",
                    "notBefore": "2021-12-02T11:35:54.000Z",
                    "serialNumber": "<snip>"
                    "subject": "/CN=<snip>",
                    "subjectAltNames": [
                        "DNS:<snip>"
                    ]
                }
            },
            "serverCertificateEncoded": "<snip>"
        },
        "port": 10443,
        "secure": true
    }
}
```

Seems like the module works like this:

* **Get list of all gateway port configurations** (i.e. load balancer endpoint configurations)
* **Get only a list of used ports** (literally ports, gets only 10443 because it is the only port that is used in the test scenario, i.e. the list that the module creates consists of a single element)
* **If port already exists then get gateway ID and get the gateway port server configs** (a foor loop enumerates the list that consists of a single element, runs a single iteration because of it and only gets the ID of the first load balancer endpoint out of the two endpoints that the alphabetically sorted JSON data includes)
* The first load balancer endpoint in terms of an alphabetical order is **bar**, so the module only takes it into account and ignores the rest of the endpoints, such as **foo**, despite the Ansible task above requires otherwise.

**Comment by joshedmonds** [comment]
@Harppi prior to StorageGRID 11.5 Load Balancer Endpoints were global only. i.e the port number was a globally unique key and the behaviour you describe was intentional due to this limitation of the Load Balancer Endpoint feature.

I'll raise an internal feature request to uplift the na_sg_grid_gateway module to support the additional endpoint binding modes introduced in SG 11.5.

**Comment by Harppi** [comment]
Thanks for the reply. I've followed up the discussion with our vendor.

**Comment by joshedmonds** [comment]
Resolved in 660542fec43b0542bd3fe3839e436f9736430884
Released as 21.9.0 on Ansible Galaxy
